### PR TITLE
Add helper class for serialization of simple types

### PR DESCRIPTION
**Pull request by ashu-mehra**
Added TR_Serializer to help serialization and deserialization of
simple data types.
Also added code for (de-)serialization of TR_BitVector.

Signed-off-by: Ashutosh Mehra <asmehra@redhat.com>

### CODE REVIEW

**Comment by ashu-mehra** [comment]
Changes in this PR are to support serialization of `TR_PersistentProfieInfo` in OpenJ9.
The OpenJ9 PR dependent on this is https://github.com/eclipse/openj9/pull/9590

**Comment by 0xdaryl** [comment]
Could you add some Doxygen comments to this new class please?

**Comment by andrewcraik** [comment]
Is there a design for how this should work at a high level explaining the API design choices? I think this would be valuable to facilitate review and for those looking to utilize this infrastructure in the future.

**Comment by vijaysun-omr** [comment]
Ping to see if this can be moved forward in the next few days

**Comment by vijaysun-omr** [comment]
@andrewcraik some new changes including comments to facilitate review have been added. So this may be ready for your review again.

**Comment by vijaysun-omr** [comment]
@andrewcraik or @ashu-mehra any update on this ?

**Comment by andrewcraik** [comment]
The details are a bit more clear - what is the purpose of this and how is it going to be used?

**Comment by ashu-mehra** [comment]
> The details are a bit more clear - what is the purpose of this and how is it going to be used?

I have created a design https://github.com/eclipse/openj9/issues/9974 for all the related changes which also shows how the `TR_Serializer` class can be used.

**Comment by vijaysun-omr** [comment]
@dsouzai fyi and @andrewcraik a ping to review the design document that was linked.

File: compiler/infra/Serializer.hpp [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+   /**
+    * @brief adds the size of the parameter to the buffer size
+    * @tparam T type of the parameter
+    * @param [in] t size of this parameter is added to the buffer size
+    * @return void
+    */
+   template<typename T>
+   void addSize(T t)
+      {
+      static_assert(std::is_trivially_copyable<T>::value, "Cannot determine size of non-trivially copyable types");
```
Comment (dsouzai) [inline compiler/infra/Serializer.hpp:59]: is `is_trivially_copyable` supported on all compiler version? I imagine with the move to the newer version of GCC and MSVC, those are probably fine, but not sure about XLC considering it was introduced in c++11.

File: compiler/infra/Serializer.hpp [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+   /**
+    * @brief adds the size of one or more parameters to the buffer size
+    * @tparam T type of the first parameter
+    * @tparam Args template parameter pack
+    * @param [in] t size of this parameter is added to the buffer size
+    * @param [in] args function parameter pack
+    * @return void
+    */
+   template<typename T, typename... Args>
```
Comment (dsouzai) [inline compiler/infra/Serializer.hpp:71]: Same point as above about variadic templates; it's listed as unsupported in [1] but it states MSVC, so maybe it's not a problem on XLC?

[1] https://github.com/eclipse/omr/blob/master/doc/SupportedC%2B%2BFeatures.md#minimum-compiler-support

File: compiler/infra/Serializer.hpp [hunk 3]
```diff
@@ -0,0 +1,281 @@
+/*******************************************************************************
+ * Copyright (c) 2000, 2019 IBM Corp. and others
```
Comment (dsouzai) [inline compiler/infra/Serializer.hpp:2]: Copyright.

File: compiler/infra/Serializer.hpp [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+    */
+   bool nextDataNotNullPointer()
+      {
+      uintptr_t ptr = *(uintptr_t *)_current;
+      TR_ASSERT_FATAL((_current + sizeof(ptr) <= _bufferEnd), "Attempting to read beyond the buffer size, _current=%p, size=%zu, _bufferEnd=%p\n", _current, sizeof(ptr), _bufferEnd);
+      _current += sizeof(ptr);
+      if (ptr)
+         {
+         return true;
+	 }
```
Comment (dsouzai) [inline compiler/infra/Serializer.hpp:222]: formatting

File: compiler/infra/BitVector.hpp [hunk 5]
```diff
@@ -163,6 +164,24 @@ class TR_BitVector
...
 
+   /**
+    * @brief Constructor to create a new BitVector by reading data from the serializer
+    * @param [in] serializer used for reading BitVector data
+    */
+   TR_BitVector(TR_Serializer &serializer)
+      {
+      _firstChunkWithNonZero = serializer.read<int32_t>();
+      _lastChunkWithNonZero = serializer.read<int32_t>();
+      _numChunks = serializer.read<int32_t>();
```
Comment (dsouzai) [inline compiler/infra/BitVector.hpp:175]: perhaps you should use the multi param version of `read` to ensure that the order in which these fields are read does not change.

**Review by dsouzai — CHANGES_REQUESTED** [review]
Overall looks ok; suggested a couple of changes, and had a couple of questions regarding the support of the c++ features used.

**Review by Leonardo2718 — CHANGES_REQUESTED** [review]
A few initial comments:

- Is this really something that the OMR project needs? The link to the OpenJ9 issues shows how this class is inteneded to be used in OpenJ9, but an explenation of the benifit this class brings to OMR is still needed. Also, if this class is to live in OMR, then the documentation explaining how it works and how to use should as well.
- If this is something the OMR project needs, why not use an exisiting C++ serialization framework (e.g. Boost Serialization)?
- If there is a reason for OMR to have its own serialization utility, then can it be made accessable/usable by other OMR components and not just the compiler?
- It looks to me like this new class is really providing three things: serialization, deserialization, and some utilities to help calculate the size of the buffer needed to serialize some data. Each of these should really be a separate class.
- We're going to need unit tests for these new utilities.

**Comment by ashu-mehra** [comment]
> Is this really something that the OMR project needs? The link to the OpenJ9 issues shows how this class is inteneded to be 
> used in OpenJ9, but an explenation of the benifit this class brings to OMR is still needed. Also, if this class is to live in OMR, then the documentation explaining how it works and how to use should as well.

It depends on the data structures we are trying to serialize. For this particular case I need to serialize `TR_BitVector` which resides in OMR. It is possible we may need to serialize other data structures as well, but as of now TR_BitVector is the only one from OMR, rest are in OpenJ9.

> If this is something the OMR project needs, why not use an exisiting C++ serialization framework (e.g. Boost Serialization)?
> If there is a reason for OMR to have its own serialization utility, then can it be made accessable/usable by other OMR components and not just the compiler?

Requirement is not from OMR project as such, but from OpenJ9. AFAIK there isn't any widespread use of serialization in the project, so I am not sure it would be a good idea to add dependency on boost library.
Others more familiar with the project code base may know better if we need OMR to have its own serialization utility. If that's the case then I agree looking at the boost framework should be the first step.

I can understand the concerns around the utility of this class for OMR project given a very narrow use-case in the project. It was only meant to be used as a helper class to avoid boiler plate code involved in reading and writing data to memory buffer. I can instead move the functionality to `TR_BitVector::serialize()` and `TR_BitVector::deserialize()` methods and get rid of the helper class completely.

**Comment by ashu-mehra** [comment]
@Leonardo2718 @dsouzai I have removed `TR_Serializer` from the change set. New changes only perform serialization of `TR_BitVector` as required by https://github.com/eclipse/openj9/pull/9590.
Removing `TR_Serializer` should address many of the concerns raised during the review.

**Comment by ashu-mehra** [comment]
This PR's main purpose was to introduce `TR_Serializer`, but now that it has been removed and is only performing (de-)serialization of `TR_BitVector`, it would be better to close this PR and open a new one.

**Comment by ashu-mehra** [comment]
I have opened https://github.com/eclipse/omr/pull/5693 which just performs (de-)serialization of `TR_BitVector`.

**Pull request closed without merging**